Adding costing to ED model

aladshaw3 opened the pull request:
## Fixes/Resolves:

Partially addresses model enhancements in #489 

## Summary/Motivation:

Adds costing methods and variables for electrodialysis to the WaterTAP costing block

## Changes proposed in this PR:
- Adding costing for electrodialysis to WaterTAP costing block
- Adding unit tests for costing in the electrodialysis unit model 
- Minor fixes to electrodialysis model 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

dangunter committed:
first changes

dangunter committed:
fixed default display/desc, added export_variables func

dangunter committed:
Example UI wrapper for METAB flowsheet

dangunter committed:
Export selected variables for UI

dangunter committed:
added driver

dangunter committed:
added build/solve

dangunter committed:
Modified display functions to take flowsheet not parent model (which they never used)

dangunter committed:
Display results command

dangunter committed:
Added deferred addition of flowsheet and action dependencies

dangunter committed:
Added update command and use of deferred flowsheet block

dangunter committed:
Some tests

dangunter committed:
Accept flowsheetInterface for load()

dangunter committed:
tweaks

dangunter committed:
special-case for build action

dangunter committed:
better handling of attempt to export non-variables.

dangunter committed:
add costing in build; driver enhancements

dangunter committed:
added "category", for costing

dangunter committed:
using category

dangunter committed:
cleanup and fix tests

dangunter committed:
start real docs

dangunter committed:
Merge remote-tracking branch 'upstream/main' into ui_api-issue524

dangunter committed:
docs for export_variables, new method add_action_type

dangunter committed:
wip docs for API

dangunter committed:
RST fix

dangunter committed:
more docs

dangunter committed:
more docs

dangunter committed:
Added custom action type and tests

dangunter committed:
Turn off the noindex in apidocs, also local convenience env var

dangunter committed:
reverted conf, more doc tweaks

dangunter committed:
Read-only load/save implemented

dangunter committed:
search / ref icon

dangunter committed:
reworded action deps, added more to-do

dangunter committed:
space for schema

dangunter committed:
generate schema, added descriptions

dangunter committed:
Converting to pydantic 1/387

dangunter committed:
Converting to pydantic 2/387

dangunter committed:
Converting to pydantic 3/387

dangunter committed:
Converting to pydantic 4/387

dangunter committed:
work save

dangunter committed:
fixed update

dangunter committed:
fix indent

dangunter committed:
thuoght experiment with flowsheet directory

dangunter committed:
test pass again

dangunter committed:
test pass again

dangunter committed:
added parameters

dangunter committed:
added discovery of flowsheet interfaces

dangunter committed:
docs updates

dangunter committed:
exported some model.Block properties as class properties

dangunter committed:
tests for model.Block properties as BlockInterface properties

dangunter committed:
improved description

Commit:
Made first attempt at adding costing to ed

aladshaw3 committed:
Merge branch 'main' into ed0d-costing

dangunter committed:
added pydantic

dangunter committed:
fixes for metab and testing of metab

dangunter committed:
Merge remote-tracking branch 'upstream/main' into ui_api-issue524

dangunter committed:
fixing tests

dangunter committed:
added a .dict() for BlockInterface

aladshaw3 committed:
Merge branch 'main' into ed0d-costing

dangunter committed:
nobody likes deprecation warnings

dangunter committed:
ran black

dangunter committed:
black fmt

dangunter committed:
Merge branch 'main' into ui_api-issue524

dangunter committed:
Merge remote-tracking branch 'upstream/main' into ui_api-issue524

dangunter committed:
remove driver code from coverage stats

dangunter committed:
Merge branch 'ui_api-issue524' of https://github.com/dangunter/watertap into ui_api-issue524

dangunter committed:
moved methods from flowsheetinterface up to superclass blockinterface

dangunter committed:
flatten_tree plus more tests

aladshaw3 committed:
Merge branch 'main' into ed0d-costing

Commit:
Finished first version of costing. Needs additional tests.

Commit:
Finalized costing with new test

Commit:
Changing a line to try to force codecov to report status.

aladshaw3 committed:
Merge branch 'watertap-org:main' into ed0d-costing

aladshaw3 committed:
Merge branch 'watertap-org:main' into ed0d-costing

Commit:
Fixed error caused by renaming of function

dangunter committed:
Merge remote-tracking branch 'upstream/main' into ui_api-issue524

dangunter committed:
Update watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/metab/tests/test_metab_api.py

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update watertap/ui/api.py

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update watertap/ui/api_model.py

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update watertap/ui/api_util.py

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update docs/how_to_guides/how_to_use_ui_api.rst

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update docs/how_to_guides/how_to_use_ui_api.rst

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update docs/how_to_guides/how_to_use_ui_api.rst

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update docs/how_to_guides/how_to_use_ui_api.rst

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update docs/how_to_guides/how_to_use_ui_api.rst

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update docs/how_to_guides/how_to_use_ui_api.rst

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update docs/how_to_guides/how_to_use_ui_api.rst

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/metab/metab_api.py

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
Update watertap/ui/api.py

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

dangunter committed:
doc improvements

dangunter committed:
fixed top-block (again) and also more robustness

dangunter committed:
Merge branch 'ui_api-issue524' of https://github.com/dangunter/watertap into ui_api-issue524

bknueven commented on watertap/costing/watertap_costing_package.py:
```diff
@@ -793,6 +868,61 @@ def cost_membrane(blk, membrane_cost, factor_membrane_replacement):
+                                        [fraction of electrodes replaced/year]
+
+    """
+
```
If this function were to be used as a stand-alone, *it* should call `make_capital_cost_var` and `make_fixed_operating_cost_var`, as opposed to requiring these vars to be created first.

bknueven replied:
```suggestion
    make_capital_cost_var(blk)
    make_fixed_operating_cost_var(blk)

```

bknueven commented on watertap/costing/watertap_costing_package.py:
```diff
@@ -658,6 +690,48 @@ def cost_caoh2_mixer(blk, dosing_rate):
+            )
+
+        make_capital_cost_var(blk)
+        make_fixed_operating_cost_var(blk)
```
```suggestion
```

bknueven commented on watertap/costing/watertap_costing_package.py:
```diff
@@ -59,6 +60,11 @@ class MixerType(StrEnum):
 
+# # TODO: Revisit types of costs for different options
+class ElectrodialysisCostType(StrEnum):
+    default = "default"
```
Do we anticipate creating more things here? Otherwise it is an unnecessary complication.

aladshaw3 replied:
I wasn't sure at the time. There are different operational modes for electrodialysis (constant current vs constant voltage). I am not sure whether or not that would change anything in the costing. May be best just to remove this for now.

bknueven reviewed (commented):
I have just a few minor comments.

lbibl commented on watertap/costing/watertap_costing_package.py:
```diff
@@ -197,6 +203,32 @@ def build_global_params(self):
+            initialize=86,
+            doc="Electrodialysis membrane cost per cell pair is sum of costs for CEM and AEM",
+            units=self.base_currency / (pyo.units.meter**2),
+        )
```
We discussed this.  But with a second thought I recommend more separating cem and aem.  The ion exchange membranes can be the same membrane with other units such as water electrolyzer. I am ok with keeping it as is.

aladshaw3 replied:
With regards to the 'specific energy calculation', I thought we had discussed this during our last chem team meeting and I thought that this was something that was automated by the costing package (i.e., the package can automatically get an energy cost per water produced). Maybe I am wrong on that. 

We can focus on getting your 1D PR merged first, then this can come after and I can make the necessary updates.

lbibl reviewed (commented):
No major issue on the costing part found by me.   Consider adding the specific energy calculation (Austin might be in progress of this).     Another thing caused by me-  I caught a minor error in the 0D unit model, which slightly changed the costing outcome (<1$).  Also I changed the name of the 0D model from "0d" to "0D" just to look better along with others .  I included these 0D-related changes in my 1D mode PR (#571) , which should be about ready to go.   I have integrated these 0D changes to the costing branch on my local and it is ready to be pushed into this PR.  I am holding off pushing this until I speak to Austin.

dangunter committed:
black

dangunter committed:
experimental additional flowsheet interface

dangunter committed:
typing and minor tweak

dangunter committed:
fixed tests, refactored simple test flowsheet

aladshaw3 committed:
Merge branch 'main' into ui_api-issue524

aladshaw3 committed:
Merge branch 'watertap-org:main' into ed0d-costing

aladshaw3 committed:
Merge branch 'watertap-org:main' into ed0d-costing

aladshaw3 commented:
Waiting on #571

aladshaw3 committed:
Update watertap/costing/watertap_costing_package.py

Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>

aladshaw3 committed:
Update watertap/costing/watertap_costing_package.py

Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>

Commit:
Moved some helper functions per Ben's suggestion

Commit:
Merge branch 'ed0d-costing' of https://github.com/aladshaw3/watertap into ed0d-costing
Merging

dangunter committed:
fixed missed call to initialize

dangunter committed:
Merge remote-tracking branch 'upstream/main' into ui_api-issue524

dangunter committed:
Merge branch 'ui_api-issue524' of https://github.com/dangunter/watertap into ui_api-issue524

Commit:
Merge branch 'main' into ed0d-costing

aladshaw3 committed:
Merge branch 'main' into ui_api-issue524

aladshaw3 committed:
Update docs/technical_reference/ui/index.rst

Changing 'automodule' to follow new doc standards

aladshaw3 committed:
Apply suggestions from code review

Removing `auto` function and classes per new doc standards.

aladshaw3 committed:
Merge branch 'watertap-org:main' into ed0d-costing

adam-a-a committed:
Refining DSPMDE  (#558)

* uncomment solve in initialization to troubleshoot

* Adding constraint scaling to all unscaled constraints but commenting some out due to worsening of solve;comment out pore vol flow constraint

* entering the rabbit hole

* refining initialization, scaling, and removing degenerate constraints

* save changes that had little/no impact on dilemma

* update bounds on velocity

* adding sim file; commented out more potential degenerate constraints; update lower bounds on couple NF variables to 0

* update more lower bounds to 0

* add sim file to unit model folder for easier access

* update sim file and some lower bounds on vars

* added expression for observed rejection for comparison with intrinsic when deactivating constraints where I suspect the problem to be hiding

* fix comment in sim file

* Commented out bounds for fluxes and Peclet number; added experimental constraints (deactivating one of them in sim file)

* delete sim file from sandbox and only keep sim file in unit models folder

* Model is working! Added electroneutrality for bulk feed to make prob square, rejection rates are as expected;
initialization and scaling still have room for improvement. Also need to enable concentration polarization again now
that the mode where CP is neglected works.

* revise scaling with conc pol disabled to improve solve

* added constraint link vol flow of pore to permeate; deactivating feed electroneutrality con in sim file

* NF model is functional with and without concentration polarization for Na and Cl! 3-5 iterations to solve

* Na, Cl, Mg working without CP; 3-5 iterations for final solve

* Na, Cl, Mg, SO4 working with CP

* All 5 ions working with CP! Na, Cl, SO4, Ca, and Mg; 3 iterations in final solve

* NF working with all ions with and without CP

* add cascading scale factors for salt fluxes instead of fixed scaling factor for all ions

* update masstransfercoeff enum default to spiral wound; remove try/except in sim file

* cleanup model and remove deactivated constraint; explicitly include CP as enumerated class

* fully incorporated conc pol config option and added mass tran coeff none type

* tests passing for 5 ions with CP calculated and spiral wound corr

* adjust ion prop test

* delete sim file

* add momentum balance type equal none

* remove bounds on electric potential and set ub of recovery to 1

* tighten outdated regression tests

* add tests without conc pol

* add tests for 2 ions; with and without CP

* add test with different inlet conditions (double concentration of each ion)

* add inverse solve test (still initializing with fixed inlet though then unfixing inlet and fixing outlet)

* add test to check mass transfer coeff fixed solve, check config errors for mass transfer with CP options

* slight cleanup and tweaks in initialization/scaling

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>

lbibl committed:
Electrodialysis_1D unit model  (#571)

* ED1d initial version done

* init_scaling added

* save file, all comp in

* save file

* resolved 1d mass-balance

* power calculation addressed

* performance metrics added

* scalling complement

* vigorously tested model, performance finalized, scaling refined

* 1d test file initiated and first version completed

* save unit model working for the frist test file

* scaling amend for 1d

* ED0d doc amend

* ED0d model DSPMDE pp, slight amendment

* black all touched files

* 0D file rename, cap D

* Capitalize d to D in ED 0d files

* black files

* 0D documentation formatting

* missed scaling amended

* tightening tol

* addressed comments from @adam-a-a

* Update watertap/unit_models/electrodialysis_1D.py

Co-authored-by: Adam Atia <aatia@keylogic.com>

Co-authored-by: Austin Ladshaw <aladshaw3@outlook.com>
Co-authored-by: Adam Atia <aatia@keylogic.com>

arao53 committed:
Dye desalination: merging sweeps for two flowsheets (#578)

* file to work in

* save some work

* API basics sketched out

* Started test of API vs. seawater_RO_desalination use case

* save work

* can run the workflow (not verifying correctness)

* Saving work

* Saving

* Saving work

* Put utility functions here

* simplified workflow and merged into FlowsheetInterface

* Added draft load/save functions

* Added docstring to as_dict()

* redoing tests

* Added more tests

* More tests and TODO

* Save work

* saving work

* Save work

* Save work again

* Save work once more

* saving work de novo

* archiving effort

* salvando o trabalho

* remarkably, the tests pass

* Schema updates and tests

* tests working

* Added some basic docs

* Added some basic docs 2

* removed outdated test file and ran Black

* docstring changes, and added some generated apidocs

* fixed apidocs

* Metab analysis (#505)

* Initial Sensitivity analyses incorporated for METAB  (#486)

* added bioreactor_simple_reaction

* add initial energy consumption to bioreactor

* addressed merge conflicts

* updated bioreactor for electricity and thermal consumption

* renamed bioreactor_simple_reaction to metab

* missed renaming test file for metab

* added tests for H2 and CH4 metab

* added costing to metab

* added costing tests for metab

* Update watertap/core/zero_order_costing.py

Co-authored-by: Adam Atia <aatia@keylogic.com>

* ran black

* addressed comments

* addressed test failure

* addressed test failures round 2

* ran black

* skip report tests on macs

* deleted demo_evaporator.py

* added flowsheet and parameter sweep for metab

* fix parameter sweep issue in write_to_csv

* add constraint for bead replacement factor equality between h2 and ch4; update replacement factor default for ch4 in yaml

* revise metab test values for fixed operating costs

* added initial set of sensitivity analyses; made generatio_ratio a variable instead of param in sido_r

Co-authored-by: TimBartholomew <timothy.bartholomew@netl.doe.gov>
Co-authored-by: TimBartholomew <timvbart@gmail.com>

* Metab flowsheet 2 (#493)

* added bioreactor_simple_reaction

* add initial energy consumption to bioreactor

* addressed merge conflicts

* updated bioreactor for electricity and thermal consumption

* renamed bioreactor_simple_reaction to metab

* missed renaming test file for metab

* added tests for H2 and CH4 metab

* added costing to metab

* added costing tests for metab

* Update watertap/core/zero_order_costing.py

Co-authored-by: Adam Atia <aatia@keylogic.com>

* ran black

* addressed comments

* addressed test failure

* addressed test failures round 2

* ran black

* skip report tests on macs

* deleted demo_evaporator.py

* added flowsheet and parameter sweep for metab

* fix parameter sweep issue in write_to_csv

* add constraint for bead replacement factor equality between h2 and ch4; update replacement factor default for ch4 in yaml

* revise metab test values for fixed operating costs

* added initial set of sensitivity analyses; made generatio_ratio a variable instead of param in sido_r

* Add METAB reactor ZO model (#484)

* added bioreactor_simple_reaction

* add initial energy consumption to bioreactor

* addressed merge conflicts

* updated bioreactor for electricity and thermal consumption

* renamed bioreactor_simple_reaction to metab

* missed renaming test file for metab

* added tests for H2 and CH4 metab

* added costing to metab

* added costing tests for metab

* Update watertap/core/zero_order_costing.py

Co-authored-by: Adam Atia <aatia@keylogic.com>

* ran black

* addressed comments

* addressed test failure

* addressed test failures round 2

* ran black

* skip report tests on macs

Co-authored-by: Adam Atia <aatia@keylogic.com>

* simplifying parameter sweep API

* updating documentation

* fixing docstring

* fixing how to monte carlo

* updated flowsheet

* ran black

* Cleaning up docstrings (#491)

* add sensitivity analyses for h2 and ch4 price

* remove pass

Co-authored-by: Adam Atia <aatia@keylogic.com>
Co-authored-by: Bernard Knueven <Bernard.Knueven@nrel.gov>
Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>

* NFC: black

* Update watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/metab/metab.py

* Update watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/metab/metab.py

* Update watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/metab/metab.py

add relevant todo note

* run black

* added test for metab flowsheet

* added multi-sweep test

* ran black

* addressed comments

* addressed test failure

* Update watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/metab/tests/test_multi_sweep.py

Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>

* cleanup test

* covering main

Co-authored-by: Adam Atia <aatia@keylogic.com>
Co-authored-by: Bernard Knueven <Bernard.Knueven@nrel.gov>
Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>
Co-authored-by: Andrew Lee <andrew.lee@netl.doe.gov>

* Costing flows in the WaterTAP costing package methods (#504)

* Minor refactoring of the parameter sweep input parser and addition of new sampling types (#522)

* fixed docstring indentation issues

* Enable recursive parameter sweep to attempt to provide a guaranteed number of solutions (#519)

* readded files pertaining to recursive parameter sweep.

* adjust failure rate and remaining sample calculation

* adding safe limits and predictions to iteration

* updating recursive param sweep test

* updating filename call signatures

* formatting new files with black

* removing deprecated code.

* use buffered mpi function and add test for parallel execution

* fixing black formatting

* clean up and addition of pragma no covers.

* reran black on recursive_param_sweep.py

* removed unused imports and minor cleanup.

* New limiting mechanism for increasing sample size in recursive iterations.

Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>

* minor cleanup and chaning how optimal indices are identified.

* ran black and removed itertools import.

* removed np.set_printoptions.

Co-authored-by: Kinshuk Panda <kinshuk.panda@nrel.gov>
Co-authored-by: Ethan Young <ethan.young@nrel.gov>
Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>

* Configure Sphinx to generate apidoc files automatically (#526)

* Configure Sphinx to generate apidoc files automatically on sphinx-build

* Remove apidoc files that were checked into Git

* Address Sphinx warnings

* Address Sphinx warning, 2nd round

* Remove scripts and code used to run sphinx-apidoc manually

* Update documentation documentation

* Modify MABR model for AMO project (#530)

* Add low pressure pump model for AMO

* Resolve merge conflict

* Resolve merge conflict

* Modify mabr model

* Resolve Mac setup problem

* Resolve solving for mac

* Change byproduct solution test

* Revise test file

* Add test for key reactant

* Rename blower size to reactor area

* Black test file

* Add reactor size to report

* Add dye_desalination.py file and import libraries.

* Added the build function to dye_desalination.py

* Updated build() in dye_desalination by adding arc connections and default scaling factors.

* Added set_operating_conditions() to dye_desalination.py

* Added specific parameters to set_operating_conditions() to dye_desalination(). Also included initialize system and solve from metab.py

* Added display function to dye_desalination.py

* Created dye_sweep.py and __init__.py to dye_desalination.

* Added skeleton to dye_sweep.py. Costing needs to be implemented and sweeps to be defined.

* Added parameter sweep in dye_sweep.py

* Created .yaml file for dye_desalination costing model

* Added TODO comments for costing model and sweep conditions.

* Added skeleton code for dye_desalination.py for the costing model and cost display

* Added default wastewater recovery cost parameters to dye_desalination_global_costing.yaml

* More descriptive comments in dye_desalination.py and dye_sweep.py

* Modified cost parameters in dye_desalination_global_costing.yaml and added costing comments to dye_desalination.py

* Added feed pump to build()

* Added constraint for pump applied pressure and NF applied pressure

* Fixed pump applied pressure (equal to nanofiltration operating pressure) and unfixed pump head

* Added initial cost metrics:
- LCODS - levelized cost of dye separation.
- Annual dye removal
- Cost of dye disposal
Relies on dye disposal unit cost ($/m3) which needs to be changed.

* Implemented annual dye sepration and dye removal cost

* Added display function for costing

* add UnitModelCostingBlocks for NF and Pump; fix variable operating cost for NF in zero order costing package

* Added levelized cost of dye separation on a per unit volume of treated feed water basis

* Added the levelized cost of dye separation and reformatted display_costing()

* Adjusted the levelized cost of water to include disposal costs based on permeate dye mass

* Added comments and documentation for dye disposal and levelized costs

* Updated sweep with case for dye cost

* Resolved issue with NaN outputs by changing optimizer function and optimization kwargs

* Changed the levelized cost of water metric to reflect dye recovery

* Added sweeps for water permeability and dye selectivity

* Added additional parameter sweeps

* Added disposal and recovery costs to print with display_costing()

* Added levelized cost of treatment (LCOT) on a dye mass basis as a new metric.

* Added a new file for dye_desalination with a reverse osmosis polish for NF permeate

* Added build function with erd types: "pressure exchanger" and "pump as turbine"

* Delete dye_desalination_RO

* Renamed variable dye_cost to dye_retentate_cost to more accurately describe costing method.

* Added test files to dye desalination

* Moved initialize system function call to the set_operating_conditons function

* Added tests for the dye desalination flowsheet

* Added the test file for the sweep tool of dye desalination

* Added tests for dye desalination sweep tool

* Added copyright header to dye sweep

* Changed display_units function to remove permeate and retentate and add P1.

* Removed feed from display results.

* Moved initialize system to main function and added assert degrees of freedom after initialize costing

* Adjusted main function call to accept new inputs

* Added initialize to set_operating conditions test.

* Added flowsheet for dye desalination with reverse osmosis unit for NF permeate treatment

* Added checks for DOF in build and set operating conditions

* Added costing method inspired by seawater_RO_desalination.py

* Resolved DOF error in set operating conditions by fixing temperature and pressure at the translator block.

* Resolved errors with RO block in initialize_system

* Save work on initialize

* Saving changes to initialize

* Resolved initialization by fixing velocity and recovery parameters instead of membrane width and area.

* Added initial costing and display

* Added costing to include externalities with dye and water recovery and brine waste.

* Differentiated between levelized cost of water (permeate) and cost of treatment (feed)

* Added secondary treatment before nanofiltration components

* Add test file and displays to dye_desalination_withRO

* Added tests for dye desalination with RO

* Additional test coverage for dye_desalination_withRO

* Fixed test values

* Removed pump_as_turbine configuration and set erd_type to pressure_exchanger always.

* Resolved discrepencies between costing for dye_desalination and dye_desalination_withRO

* Resolved test errors due to costing and naming convention changes

* Resolved differences between sweeps for dye_desalination and dye_desalination_withRO

* Added integration tests for dye_sweep.py

* Updated a few leftover cost units to USD2020

* Update watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/dye_desalination/dye_sweep.py

Co-authored-by: Dan Gunter <dkgunter@lbl.gov>
Co-authored-by: TimBartholomew <timothy.bartholomew@netl.doe.gov>
Co-authored-by: Adam Atia <aatia@keylogic.com>
Co-authored-by: Bernard Knueven <Bernard.Knueven@nrel.gov>
Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>
Co-authored-by: Andrew Lee <andrew.lee@netl.doe.gov>
Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>
Co-authored-by: Kinshuk Panda <kinshuk.panda@nrel.gov>
Co-authored-by: Ethan Young <ethan.young@nrel.gov>
Co-authored-by: Ludovico Bianchi <lbianchi@lbl.gov>
Co-authored-by: luohezhiming <98901358+luohezhiming@users.noreply.github.com>

dangunter committed:
Merge branch 'main' into ui_api-issue524

dangunter committed:
Merge pull request #527 from dangunter/ui_api-issue524

This is the API we will use to wrap flowsheets for the UI demo on 6/28. Needs to get merged ASAP.

Commit:
Merge conflicts

Commit:
Updated test values

Commit:
Merge branch 'aladshaw3-ed0d-costing'

aladshaw3 commented:
Dealing with merge conflicts right now. May have to open a new PR.

aladshaw3 commented:
Merge conflict resolution led to very strange tracking of changes. Replacing this PR with #582
